### DISCUSSION TITLE
Extend prPriority features to branches

### DISCUSSION THREAD

**Discussion by jhaumont** [renovatebot/renovate#19641] [Suggest an Idea]
### What would you like Renovate to be able to do?

Hello! Happy new year 😄 

`prPriority` is a good feature. But it's not enough for us (in our context). We are very interested by the possibility to have a `branchPriority` too. Let me explain.

We have a flux repository where we handle all of our apps deployment as GitOps. You have 5 differents env in the same repo. There is an order (from the less critical to the most) and they are classified by level.

dev -> preprod -> prod-1, prod-2, prod-3

Here is our configuration with `prPriority` setup by level:
```json5
// dev level
{
  "prPriority": 50,
  "matchPaths": [
     "dev/**",
  ],
},
// preprod level
{
  "prPriority": 25,
  "matchPaths": [
     "preprod/**",
  ],
},
// prod level
{
  "prPriority": 0, // default value
  "matchPaths": [
     "prod-1/**",
     "prod-2/**",
     "prod-3/**",
  ],
},
```

But, with this, only PR follow the order. If branches are not created in the correct order (because we have reach our `branchConcurrentLimit` for example), this is a non-sense for us.
We could have a PR for prod workload before having a PR for dev & preprod.

Thanks a lot for the answer and maybe for the future PR 🙏 

### If you have any ideas on how this should be implemented, please tell us here.

No sorry

### Is this a feature you are interested in implementing yourself?

No, I'm not confortable at all in JS so, not available to contribute :/

**Comment by rarkins** [comment]
I have converted this to a Discussion, because I'm not sure it' actionable. I'm not actually sure what you mean by this, so please describe in more details.

**Reply by usmonster** [reply]
Hello! Thanks for your prompt response. I'll permit myself to restate the problem & feature request (@jhaumont, please correct me if I misspeak):

* We have an IaC repo with root-level directories reflecting different environments in our infrastructure (e.g. `preproduction`, `production`, etc.)
* Contents and dependencies are very similar (essentially mirrored) across environments, but we'd like any given dependency update to be proposed for the `preproduction` environment before `production`, so they can be tested before reaching a critical environment.
* Today, Renovate branches are generated in arbitrary order, so the dependency update branch for `production` may be created before that for `preproduction`;
* This means that, despite our `prPriority` configuration to prioritize `preproduction` updates, we might still see the dependency update PR for `production` first, which is not ideal.

It would be great if there was a clean way to do this via, for example, a `branchPriority` feature which could, for example, scan for all updates in a repo and prioritize creating branches with updates in locations matching the higher-priority patterns.

I hope this is clearer, and if so, can you please recommend any action to take (e.g. reconvert this into an issue)? Thanks again!

**Reply by rarkins** [reply]
You can write packageRules with `matchPaths` and `prPriority`. e.g. give production paths higher priority

**Reply by usmonster** [reply]
Hi @rarkins, this would not be sufficient for the case I described, as made explicit in my last bullet point.

From what I understand, `prPriority` looks at _existing_ Renovate-generated branches and uses configured priorities to choose the order in which to create PRs from those branches. We need something that looks at available upgrades in the matching paths and sets the priority order in which _branches_ should be created by Renovate, so a similar principle, but earlier in the lifecycle.

Please let me know if this still isn't clear, and thanks again for your fast response!

**Reply by rarkins** [reply]
`prPriority` sorting happens prior to branch logic. It does not happen between branches and PRs.

Ultimately, a branch won't be created before another one which has a lower priority.

I'm not sure I understand yet what you mean, and I'm not sure if you understand how Renovate works. We might both have gaps. For example you wrote "Renovate branches are generated in arbitrary order" which is incorrect.

Instead of you describing internal Renovate concepts which you might be misunderstanding, can you instead describe using an example? Use as simple and concrete of an example as you can.

e.g. branch X exists with a PR. branch Y exists with a branch but not PR.

And then describe the difference between current behavior and desired behavior.

**Reply by usmonster** [reply]
> For example you wrote "Renovate branches are generated in arbitrary order" which is incorrect.

Sorry; to clarify, I meant that it seems to me that Renovate branches aren't generated in an order that takes specified `prPriority` configuration into account, which may be the main source of misunderstanding if this is not the case. It may indeed be more of a documentation issue since this behavior is not clear from [what is written](https://docs.renovatebot.com/configuration-options/#prpriority).

In any case, we have previously observed this behavior, so I'll give a simplified concrete example as requested.

**Context:**
* `prConcurrentLimit: 1, branchConcurrentLimit: 1, prCreation: "not-pending"`
* `{"prPriority": 1, "matchPaths": ["preproduction/**"]}, {"prPriority": 0, "matchPaths": ["production/**"]}`
* No open PRs or branches from Renovate yet
* Renovate scans and detects an available update D

**Observed behavior:** branch A is created with update D for `production`
**Desired behavior:** branch B is created with update D for `preproduction`

Looking at our repository history this seems rare, but it has happened. I haven't tried to reproduce it explicitly, but this is the main issue we'd like to solve. I'll let @jhaumont take over if you'd like more information. Thanks yet again!

**Reply by rarkins** [reply]
First clarification: the naming and documentation of prPriority comes from when branch & PR were completely tied together so saying PR priority was essentially the same as branch priority. These days there are a few ways to separate them, e.g. create the branch but wait for tests to finish before creating the PR. Ultimately the logic and priority sorting happens prior to branch creation or update.

Branches will always be sorted by prPriority (if defined) prior to any create or update branch logic.

Now back to your example..

If update D is available/applicable for both `production` and `preproduction`, then the `prPriority` value should be the deciding factor and the _branch_ for `preproduction` is created first. If it didn't happen like this, it's either a bug, or perhaps something wrong with the matchPaths.

Your setting of 1 concurrent branch and prCreation=not-pending may lead to behavior you find unintuitive at times.

e.g. let's say that dependency D is created and then merged for `preproduction`, and then the branch is created for `production` (but no PR yet, because it's still pending). Then on the next run, dependency E is available for both. Even though dependency E on `preproduction` is higher priority, Renovate cannot/will not create a branch for it because there already exists the maximum 1 branches open. So it will create a PR for dependency D, and wait for that to merge before any for E can be created.

With the above config, I can't think of a scenario where either D or E would be created for `production` before `preproduction` if all else is equal. In other words `prPriority` does apply to branches, and does do what I think you need it to do.

**Reply by usmonster** [reply]
Thanks so much for your time and explanations! This definitely clarifies expected behavior and confirms some things we noticed, but we'll keep an eye out to see if we ever witness behavior contrary to what's expected.
